client/Client.hs

```diff
@@ -1,30 +1,368 @@
+withNats opts conn callback = do
+  -- apply initial opts and set puller
+  let client = newClient opts
+
+  let inbox = Buffer.newBuffer (recv conn)
+  let outbox =  []
+  let sender = send conn
+  let client' = client {inbox = inbox, outbox = outbox, sender = sender, closer = close conn}
+  tClient <- newTVarIO client'
+
+  waitForInfo tClient
+  routeMessages tClient
+
+  connect tClient
+  sendBytes tClient
+
+  callback tClient -- TODO: this will need to be async so we get to loop, but we need to know when it's done... or we make loop async, but keep track of it to cancel on close.. that would be nicer
+  sendBytes tClient
+
+  loop tClient -- TODO: ideally we want two loops, one for reading and one for writing
+
+loop :: TVar Client -> IO ()
+loop c = do
+  c' <- readTVarIO c
+  a <- newEmptyTMVarIO
+  b <- newEmptyTMVarIO
+  -- read incoming messages
+  void . forkIO $ do
+    case status c' of
+      CLOSING -> routeMessages c
+      _       -> readMessages c >> routeMessages c
+    atomically (putTMVar a ())
+  -- send outgoing messages
+  void $ forkIO (sendBytes c >> atomically (putTMVar b ()))
+  --wait for them to resolve
+  atomically $ takeTMVar a
+  atomically $ takeTMVar b
+  --decide what to do next
+  client <- readTVarIO c
+  case status client of
+    CLOSING -> cleanup c
+    _       -> loop c
+
+cleanup :: TVar Client -> IO ()
+cleanup client = do
+  client' <- readTVarIO client
+  -- if we've got nothing left to send or route then close the conn, otherwise send/route first then close
+  case flushed client' of
+    True -> closer client'
+    _    -> loop client
+
+  where
+    flushed c = out c && in' c
+    out c = null (outbox c)
+    in' c = (BS.length . Buffer.bytes . inbox $ c) == 0
+
+class NatsConn a where
+  recv :: a -> Int -> IO BS.ByteString
+  send :: a -> BS.ByteString -> IO ()
+  close :: a -> IO ()
+
+readMessages :: TVar Client -> IO ()
+readMessages client = do
+  client' <- readTVarIO client
+  inbox' <- execStateT Buffer.hydrate (inbox client')
+  atomically . modifyTVar client $ \c -> c {inbox = inbox'}
+
+waitForInfo :: TVar Client -> IO ()
+waitForInfo client = do
+  readMessages client
+  client' <- readTVarIO client
+  let bytes = Buffer.bytes . inbox $ client'
+  case BS.length bytes of
+    0 -> threadDelay 10000 >> waitForInfo client
+    _ -> return ()
+
+routeMessages :: TVar Client -> IO ()
+routeMessages client = do
+  client' <- readTVarIO client
+  let inbox' = inbox client'
+  let bs = Buffer.bytes inbox'
+  case BS.length bs of
+    0 -> threadDelay 10000
+    _ -> do
+      case genericParse bs of
+        -- TODO: alter some top level state to indicate a parse error
+        Left err -> error (show err)
+        Right (msg, rest) -> do
+          -- remove the parsed message from the buffer
+          inbox'' <- execStateT (Buffer.chomp (BS.length bs - BS.length rest)) inbox'
+          atomically . modifyTVar client $ \c -> c {inbox = inbox''}
```

Comment:
This is also questionable, it feeds in a stale version of the inbox

Reply:
If the inbox is implemented in the client this could be simplified and possibly composed with adding the pulled bytes to the inbox